File: backend/app/renewals.py
# Renewal date arithmetic, kept in one place and free of any database or
# schema imports so models.py can use it without a circular import.
#
# The one thing worth knowing before reading any of this: `timedelta` cannot
# express "one month". A month is 28, 29, 30 or 31 days depending on which one
# it is, so every function here works in whole calendar months and asks the
# calendar how long each one actually is.

from calendar import monthrange
from collections.abc import Iterator
from datetime import date, timedelta


def add_months(anchor: date, months: int) -> date:
    """`anchor` shifted forward by whole calendar months.

    A day that does not exist in the target month is clamped to that month's
    last day: the 31st of January plus one month is the 28th of February.

    The clamp is always computed from the *anchor*, never from the previous
    result, which is what keeps a plan anchored on the 31st billing on the
    31st again in March instead of being permanently dragged back to the 28th
    by one short February.
    """
    # Counting months from year zero makes the year rollover plain division
    # rather than a loop with a special case at December.
    total = (anchor.year * 12) + (anchor.month - 1) + months
    year, month = divmod(total, 12)
    return date(year, month + 1, min(anchor.day, monthrange(year, month + 1)[1]))


def next_occurrence(anchor: date, cycle_months: int, on_or_after: date) -> date:
    """The first renewal on or after `on_or_after`, for a plan that renews
    every `cycle_months` months starting from `anchor`.

    An anchor in the future is itself the next occurrence -- nothing is
    rolled, so a subscription added today with a renewal date next month keeps
    the date the client sent.
    """
    if anchor >= on_or_after:
        return anchor
    # Jump straight to roughly the right period instead of stepping a month at
    # a time: an anchor left untouched since 2020 is one calculation, not
    # seventy iterations. The month count ignores the day of the month, so the
    # estimate can land one period short -- never more -- and the loop below
    # settles it.
    elapsed_months = (on_or_after.year - anchor.year) * 12 + (on_or_after.month - anchor.month)
    periods = elapsed_months // cycle_months
    occurrence = add_months(anchor, periods * cycle_months)
    while occurrence < on_or_after:
        periods += 1
        occurrence = add_months(anchor, periods * cycle_months)
    return occurrence


def occurrence_on_or_after(anchor: date, cycle_months: int, on_or_after: date) -> date:
    """The first date in the schedule on or after `on_or_after`, counted in
    both directions from `anchor`.

    next_occurrence deliberately never looks earlier than the anchor: it
    answers "when is the next renewal", and an anchor the client sent for next
    month *is* that answer. This one answers a different question -- where
    does the repeating schedule land inside a given window -- and for that a
    window earlier than the anchor is the normal case, not a mistake. A plan
    anchored on next March has been billing on the same day every year before
    that, and the spend summary has to be able to see those charges.
    """
    elapsed_months = (on_or_after.year - anchor.year) * 12 + (on_or_after.month - anchor.month)
    # Floor division, so a window before the anchor gives a negative period
    # count. Truncation towards zero would round the wrong way there and land
    # after the window instead of before it.
    periods = elapsed_months // cycle_months
    occurrence = add_months(anchor, periods * cycle_months)
    # The month count above ignores the day of the month, so the estimate can
    # be one period out either way -- never more. These two settle it, and
    # only one of them ever runs.
    while occurrence < on_or_after:
        periods += 1
        occurrence = add_months(anchor, periods * cycle_months)
    while (earlier := add_months(anchor, (periods - 1) * cycle_months)) >= on_or_after:
        periods -= 1
        occurrence = earlier
    return occurrence


def _walk(
    anchor: date, cycle_months: int, start: date, end: date, first
) -> Iterator[date]:
    """Every date in the schedule between `start` and `end`, both ends
    included, beginning wherever `first` says the window opens.

    Each step is measured from the anchor rather than from the previous
    result, so the month-end clamping in add_months cannot accumulate.
    """
    occurrence = first(anchor, cycle_months, start)
    while occurrence <= end:
        yield occurrence
        occurrence = first(anchor, cycle_months, occurrence + timedelta(days=1))


def occurrences_between(
    anchor: date, cycle_months: int, start: date, end: date
) -> Iterator[date]:
    """Every renewal falling in `start`..`end`, both ends included.

    A monthly plan renews several times in a 90-day window, and each of those
    renewals is a separate charge, so they are yielded separately rather than
    collapsed into the first one.

    Nothing before the anchor is yielded, which is what /upcoming wants: a
    subscription whose renewal date is next month is not also due this month.
    Use charges_between for the backward-looking view.
    """
    return _walk(anchor, cycle_months, start, end, next_occurrence)


def charges_between(
    anchor: date, cycle_months: int, start: date, end: date
) -> Iterator[date]:
    """The same window, with the schedule extended backwards through the
    anchor as well as forwards -- every date the plan billed on, not only the
    ones still to come. This is what the spend summary asks for, since the
    period it is totalling up is usually already over.
    """
    return _walk(anchor, cycle_months, start, end, occurrence_on_or_after)

File: frontend/src/renewals.js
// Renewal date arithmetic in the browser, mirroring backend/app/renewals.py.
//
// Why this exists at all: GET /subscriptions/upcoming is anchored to today --
// `days` counts forward from now and tops out at 365 -- while the design's
// "Coming up" panel lists the charges in the *selected* period, and the period
// picker reaches back to January 2025 and forward to December 2027. The
// backend arithmetic is general (renewals.occurrences_between takes any start
// and end); only the route signature is today-shaped. Until it grows a
// from/to pair (TODO.md D2) the client does the same walk itself.
//
// One caveat, and it is the reason this is a mirror rather than the truth: the
// API returns `next_renewal_date`, the derived next occurrence, not the stored
// anchor. For a plan anchored on the 31st the derived date may already be
// clamped (28 Feb), and stepping from a clamped date keeps the clamp instead
// of springing back to the 31st in March, which is what the server does. Off
// by up to three days at month ends, and only for those plans.
//
// Days are handled as {y, m, d} triples rather than Date objects throughout:
// a Date built from "2026-09-04" is UTC midnight and reads as the 3rd in any
// timezone west of Greenwich, which is exactly the kind of bug a page full of
// renewal dates cannot afford.

const daysInMonth = (year, month) => new Date(year, month + 1, 0).getDate();

function parseISO(iso) {
  const [y, m, d] = iso.split("-").map(Number);
  return { y, m: m - 1, d };
}

function toISOParts({ y, m, d }) {
  return `${y}-${String(m + 1).padStart(2, "0")}-${String(d).padStart(2, "0")}`;
}

// Ordinal comparison without building Dates.
const rank = ({ y, m, d }) => y * 10000 + m * 100 + d;

// `anchor` shifted by whole calendar months, clamping a day the target month
// does not have (31 Jan + 1 month = 28 Feb). Always measured from the anchor,
// never from the previous result, so the clamp cannot accumulate.
function addMonths(anchor, months) {
  const total = anchor.y * 12 + anchor.m + months;
  const y = Math.floor(total / 12);
  const m = ((total % 12) + 12) % 12;
  return { y, m, d: Math.min(anchor.d, daysInMonth(y, m)) };
}

/**
 * Every renewal of `subscription` that falls inside the given calendar month.
 *
 * Cancelled and paused plans yield nothing -- they are not going to be billed,
 * whatever their dates say. A trial yields its conversion date at most once,
 * because a trial converts one time and then bills under a different status;
 * treating its date as a recurring anchor would invent conversions that never
 * happen. Renewals before `started_date` are skipped, so a plan that starts
 * next quarter is not shown as due this month.
 *
 * Returns [{ iso, cost, isTrialConversion }], ordered by day.
 */
export function occurrencesInMonth(subscription, year, month) {
  const status = subscription.status;
  if (status === "cancelled" || status === "paused") return [];

  const anchor = parseISO(subscription.next_renewal_date);
  const started = subscription.started_date ? parseISO(subscription.started_date) : null;
  const firstOfMonth = { y: year, m: month, d: 1 };
  const lastOfMonth = { y: year, m: month, d: daysInMonth(year, month) };

  if (status === "trial") {
    if (rank(anchor) < rank(firstOfMonth) || rank(anchor) > rank(lastOfMonth)) return [];
    // At a cost of 0: nothing leaves the account on a conversion day, the
    // trial is free until it ends. The real price travels with the row so a
    // caller can still say "then EUR 9.99/mo".
    return [{ iso: toISOParts(anchor), cost: 0, isTrialConversion: true }];
  }

  const step = subscription.billing_cycle === "yearly" ? 12 : 1;
  // Jump straight to roughly the right period rather than stepping a month at
  // a time: the selected period can be years away from the anchor.
  const elapsed = (year - anchor.y) * 12 + (month - anchor.m);
  const periods = Math.floor(elapsed / step) * step;

  const out = [];
  // Two periods of slack on either side absorbs the day-of-month rounding in
  // `elapsed`, which ignores the day and so can land one period out.
  for (let n = periods - step * 2; n <= periods + step * 2; n += step) {
    const occurrence = addMonths(anchor, n);
    if (occurrence.y !== year || occurrence.m !== month) continue;
    if (started && rank(occurrence) < rank(started)) continue;
    out.push({
      iso: toISOParts(occurrence),
      cost: Number(subscription.cost),
      isTrialConversion: false,
    });
  }
  return out;
}

/** Every charge due in one month, across every subscription, soonest first. */
export function chargesInMonth(subscriptions, year, month) {
  const rows = [];
  for (const subscription of subscriptions) {
    for (const occurrence of occurrencesInMonth(subscription, year, month)) {
      rows.push({ subscription, ...occurrence });
    }
  }
  // Name and id break ties only to keep two charges on the same day in a
  // stable order rather than whatever the list happened to arrive in.
  rows.sort(
    (a, b) =>
      a.iso.localeCompare(b.iso) ||
      a.subscription.name.toLowerCase().localeCompare(b.subscription.name.toLowerCase()) ||
      a.subscription.id - b.subscription.id,
  );
  return rows;
}

/** How many charges fall in a whole calendar year -- the yearly KPI. */
export function chargeCountInYear(subscriptions, year) {
  let count = 0;
  for (let month = 0; month < 12; month += 1) {
    count += chargesInMonth(subscriptions, year, month).length;
  }
  return count;
}
